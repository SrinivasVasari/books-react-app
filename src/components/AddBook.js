import React, { useState, useContext } from "react";
import BooksContext from "../context/context";

const AddBook = () => {
  // eslint-disable-next-line no-unused-vars
  const { state, dispatch } = useContext(BooksContext);
  const [bookName, setBookName] = useState("");
  const [bookAuthor, setBookAuthor] = useState("");
  const [bookDesc, setBookDesc] = useState("");

  const handleSubmit = (event) => {
    event.preventDefault();
    const addBook = {
      bookName,
      bookAuthor,
      bookDesc,
    };
    if (
      bookName.trim() === "" ||
      bookAuthor.trim() === "" ||
      bookDesc.trim() === ""
    ) {
      alert("Please enter the book details");
    } else {
      dispatch({ type: "ADD_BOOK", payload: addBook });
      setBookName("");
      setBookAuthor("");
      setBookDesc("");
    }
  };

  return (
    <form className="form col s12" onSubmit={handleSubmit}>
      <div className="row customRow">
        <div className="input-field col s12 m6">
          <input
            type="text"
            name="name"
            id="name"
            onChange={(e) => setBookName(e.target.value)}
            value={bookName}
            className="name"
          />
          <label htmlFor="name">Book Name</label>
        </div>
        <div className="input-field col s12 m6">
          <input
            type="text"
            name="author"
            id="author"
            onChange={(e) => setBookAuthor(e.target.value)}
            value={bookAuthor}
            className="Author"
          />
          <label htmlFor="author">Book Author</label>
        </div>
      </div>

      <div className="row customRow">
        <div className="input-field col s12 m12">
          <input
            type="text"
            name="description"
            id="description"
            onChange={(e) => setBookDesc(e.target.value)}
            value={bookDesc}
            className="description"
          />
          <label htmlFor="description">Book Description</label>
        </div>
      </div>
      <button type="submit" className="waves-effect waves-light btn-small">
        Add Book
      </button>
    </form>
  );
};
export default AddBook;
