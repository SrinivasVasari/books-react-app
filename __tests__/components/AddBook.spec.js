import React from "react";
import { shallow, mount } from "enzyme";
import toJson from "enzyme-to-json";
import AddBook from "../../src/components/AddBook";
import BooksContext from "../../src/context/context";

describe("AddBook Container", () => { 
  let wrapper;
  let alertSpy;
  beforeEach(() => {
    alertSpy = jest.spyOn(window, "alert");
    wrapper = shallow(<AddBook />);
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  it("Should render correctly", () => {
    expect(toJson(wrapper)).toMatchSnapshot();
  });
  it("Should have three input fields", () => {
    expect(wrapper.find('input[type="text"]').length).toEqual(3);
  });
  it("Should call handleSubmit function when form is submitted", () => {
    const dispatchMock = jest.fn();
    const handleSubmitFn = jest.fn();
    const fakeEvent = { preventDefault: () => {} };
    const wrapper = mount(
      <BooksContext.Provider value={{ dispatch: dispatchMock }}>
        <AddBook onSubmit={handleSubmitFn} />
      </BooksContext.Provider>
    );
    const form = wrapper.find("form");
    wrapper.find("#name").simulate("change", {
      target: { name: "name", value: "blah" },
    });
    wrapper.find("#author").simulate("change", {
      target: { name: "author", value: "blah1" },
    });
    wrapper.find("#description").simulate("change", {
      target: { name: "description", value: "blah2" },
    });
    form.simulate("submit", fakeEvent);
    expect(handleSubmitFn).toHaveBeenCalledTimes(0);
  });
  it("Should call alert inside handleSubmit function when when submitted with empty values", () => {
    const dispatchMock = jest.fn();
    const handleSubmitFn = jest.fn();
    const fakeEvent = { preventDefault: () => {} };
    const domAlert = window.alert;
    window.alert = () => {};
    const wrapper = mount(
      <BooksContext.Provider value={{ dispatch: dispatchMock }}>
        <AddBook onSubmit={handleSubmitFn} />
      </BooksContext.Provider>
    );
    const form = wrapper.find("form");
    wrapper.find("#name").simulate("change", {
      target: { name: "name", value: "" },
    });
    form.simulate("submit", fakeEvent);
    expect(handleSubmitFn.mock.calls).not.toBeNull();
    expect(alertSpy).not.toBeNull();
    window.alert = domAlert;
  });

  it("Should render submit button with Add Book text", () => {
    const wrapper = mount(<AddBook />);
    const button = wrapper.find("button");
    expect(button).toHaveLength(1);
    expect(button.prop("type")).toEqual("submit");
    expect(button.text()).toEqual("Add Book");
  });
});
