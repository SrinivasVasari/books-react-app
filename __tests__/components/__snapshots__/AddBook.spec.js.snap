// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddBook Container Should render correctly 1`] = `
<form
  className="form col s12"
  onSubmit={[Function]}
>
  <div
    className="row customRow"
  >
    <div
      className="input-field col s12 m6"
    >
      <input
        className="name"
        id="name"
        name="name"
        onChange={[Function]}
        type="text"
        value=""
      />
      <label
        htmlFor="name"
      >
        Book Name
      </label>
    </div>
    <div
      className="input-field col s12 m6"
    >
      <input
        className="Author"
        id="author"
        name="author"
        onChange={[Function]}
        type="text"
        value=""
      />
      <label
        htmlFor="author"
      >
        Book Author
      </label>
    </div>
  </div>
  <div
    className="row customRow"
  >
    <div
      className="input-field col s12 m12"
    >
      <input
        className="description"
        id="description"
        name="description"
        onChange={[Function]}
        type="text"
        value=""
      />
      <label
        htmlFor="description"
      >
        Book Description
      </label>
    </div>
  </div>
  <button
    className="waves-effect waves-light btn-small"
    type="submit"
  >
    Add Book
  </button>
</form>
`;
